PROPOSAL.md

```diff
@@ -0,0 +1,336 @@
+| Total   | 20 bytes | 0.66 sec |
+| Speedup | 244,000  | 74.76    |
+
+## Implementation suggestions
+
+### Persistence
+
+In the Client, persist the JSON-format metadata directly to SQLite, without
+further de/serialization apart from [JSON-schema validation][JSON Schema]. I've been
+accumulating the hypothesis that this will be the easiest way to consume the
+Journalist API's JSON responses from a JavaScript/TypeScript app anyway, and
+this week @legoktm sent me a [testimonial to this
+approach](https://crawshaw.io/blog/programming-with-agents) (see the section
+beginning "Example: SQL conventions around JSON").
+
+[JSON Schema]: https://json-schema.org/overview/what-is-jsonschema
+
+### Limiting response sizes[^2]
+
+This proposal seeks to avoid (a) timestamped-based cursors, which are
+challenging to implement robustly for all our queries of interest[^3]; and (b)
+pagination, which introduces new state to track during a given sync iteration.
+
+However, let's consider each of these strategies to be a form of sharding.
+Cursors shard across timestamps, and pagination shards over the size of the
+collection. What else can we shard? One option is the `Source.uuid` field,
+which should be uniformly distributed:
+
+```sh-session
+$ python sync_client.py --prefix 0
+: 1368
+$ python sync_client.py --prefix 1
+0: 80
+1: 85
+2: 95
+3: 76
+4: 79
+5: 89
+6: 85
+7: 77
+8: 80
+9: 90
+a: 93
+b: 90
+c: 94
+d: 88
+e: 80
+f: 87
+```
+
+This suggests a future refinement of this sync strategy:
+
+1. `/head`: Server returns a version per shard.
+2. Client determines which of its shards are out of date.
+3. `/index`: Client requests updated indexes for just those shards.
+
+> [!NOTE]
+> For synchronizing a more heterogeneous collection (with many types, arbitrary
+> schemas, etc.), this approach basically generalizes to a Merkle tree, at least
+> within each shard. That generalization has some advantages; in particular, it
+> would lay the groundwork for peer-to-peer sync among Workstations using the
+> SecureDrop Protocol, where the server is only a message queue. But that would be
+> overengineered for this "v2" sync strategy, where we can count on a fixed
+> two-level hierarchy between sources and their collections.
+
+### Client-to-Server writes
+
+The Journalist API currently accepts the following writes from clients:
+
+| Resource            | Verb         |
+| ------------------- | ------------ |
+| Source (account)    | `DELETE`[^4] |
+| Source conversation | `DELETE`     |
+| Star                | `POST`       |
+| Star                | `DELETE`     |
+| Reply               | `POST`       |
+| "Seen" record       | `POST`       |
+
+Instead of maintaining a "job queue" of Client-side writes—and having to retry
+and reconcile them—make them blocking. For example, block the Client on starring
+a new source until the Server accepts it; block the Client until a new reply has
+been sent; etc. Keeping user-initiated actions synchronous will eliminate both
```

Comment:
I have less context on the network constraints here, but I do agree that it's conceptually much simpler to have the writes block on server response, and treat the server state as the sole source of truth for write success.